aten/src/ATen/native/xnnpack/Pooling.cpp

```diff
@@ -0,0 +1,321 @@
+         true;
+}
+
+Tensor run(
+    const Context& context,
+    const Tensor& input) {
+  using namespace internal;
+
+  TORCH_CHECK(
+      usable(input, context.output.height, context.output.width),
+      "mobile::cpu::max_pool not usable!");
+
+  const Tensor input_nhwc = input.contiguous(MemoryFormat::ChannelsLast);
+  Tensor output_nhwc = empty_with_tail_padding(
+      {
+        input_nhwc.size(Layout::Activation4D::batch),
+        context.output.channels,
+        shape(input_nhwc.size(Layout::Activation4D::height), context.output.height),
+        shape(input_nhwc.size(Layout::Activation4D::width), context.output.width),
+      },
+      input_nhwc.options().dtype(),
+      MemoryFormat::ChannelsLast);
+
+  const xnn_status setup_status = xnn_setup_max_pooling2d_nhwc_f32(
+      context.max_pool_op.get(),                      // operator
+      input_nhwc.size(Layout::Activation4D::batch),   // batch_size
+      input_nhwc.size(Layout::Activation4D::height),  // input_height
+      input_nhwc.size(Layout::Activation4D::width),   // input_width
+      input_nhwc.data_ptr<float>(),                   // input
+      output_nhwc.data_ptr<float>(),                  // output
+      nullptr);                                       // threadpool
+
+  TORCH_CHECK(
+      xnn_status_success == setup_status,
+      "xnn_setup_max_pooling2d_nhwc_f32 failed!");
+
+  const xnn_status run_status = xnn_run_operator(
+      context.max_pool_op.get(),  // operator
+      nullptr);                   // threadpool
+
+  TORCH_INTERNAL_ASSERT(
+      xnn_status_success == run_status,
+      "xnn_run_operator failed!");
+
+  return output_nhwc.contiguous(input.suggest_memory_format());
+}
+
+Tensor create_and_run(
+    const Tensor& input,
+    const IntArrayRef kernel,
+    const IntArrayRef padding,
+    const IntArrayRef stride,
+    const IntArrayRef dilation,
+    const bool ceil_mode,
+    const float output_min,
+    const float output_max) {
+  using namespace internal;
+
+  const int64_t input_channels = input.size(Layout::Activation4D::channels);
+  const int64_t output_channels = input_channels;
+
+  return internal::max_pool2d::run(
+      internal::max_pool2d::create(
+          input_channels,
+          output_channels,
+          kernel,
+          padding,
+          stride,
+          dilation,
+          ceil_mode,
+          output_min,
+          output_max),
+      input);
+}
+
+} // namespace
+} // namespace max_pool2d
+} // namespace internal
+
+bool use_max_pool2d(
+    const Tensor& input,
+    const IntArrayRef kernel,
+    const IntArrayRef padding,
+    const IntArrayRef stride,
+    const IntArrayRef dilation,
+    const bool ceil_mode) {
+  using namespace internal;
+
+  const int64_t input_channels = (input.dim() == 4) ?
+                                 input.size(Layout::Activation4D::channels) :
+                                 -1 /* Setting to a negative number to fail avalable() */;
```

Comment:
This seems like too tight of a coupling.  Can you add a separate check for dim==4 instead?

Reply:
Done.